### ISSUE TITLE
MCMC is slow: `random.split` inside `fori_loop` is slow

### ISSUE THREAD

**Issue by PhilipVinc** [tillahoffmann/jax-mps#196]
This is a follow up on #193 . 
#193 did fix the MWE provided there, which was reduced from a MCMC code, but lead only to a minor speedup so I got back to look what's the problem.
Claude reduced the following MWE below, and the working hypothesis is that calls to some primitives of `random.split` inside of a Jax loop requires synchronization and defeats the optimization introduced in #194 .

I'm pasting the MWE and lightly edited claude diagnosis below.

 **Any `jax.random` (threefry) op inside a `lax.scan` / `lax.fori_loop` body is dispatched
 synchronously once per iteration, defeating the #194 fast path.** A loop that only does
 arithmetic runs at ~6 µs/iter; adding a *single* `jax.random.split` to the body jumps to
 ~190 µs/iter — a ~30× penalty. `JAX_MPS_ASYNC_DISPATCH=1` does **not** help.

## MWE

```python
import time, jax, jax.numpy as jnp
from jax import lax

N = 1260  # loop iterations

def us_per_iter(fn, arg):
    f = jax.jit(fn)
    jax.block_until_ready(f(arg))  # warm/compile
    t0 = time.perf_counter()
    jax.block_until_ready(f(arg))
    return (time.perf_counter() - t0) / N * 1e6

# identical loops; the only difference is one jax.random.split in the body
arith = lambda x: lax.fori_loop(0, N, lambda i, c: jnp.tanh(c) + 1e-3, x)
rng   = lambda k: lax.fori_loop(0, N, lambda i, c: jax.random.split(c)[0], k)

print(f"backend={jax.default_backend()}")
print(f"arithmetic body     : {us_per_iter(arith, jnp.ones(16)):6.1f} us/iter")
print(f"+ jax.random.split  : {us_per_iter(rng, jax.random.PRNGKey(0)):6.1f} us/iter")
```

## Results (Metal, jax-mps 0.10.8, M-series Mac)

```
backend=mps
arithmetic body     :    5.9 us/iter
+ jax.random.split  :  189.2 us/iter      # ~32x slower
```

The penalty is the same for `lax.scan` and `lax.fori_loop`, is independent of the RNG
output shape (`split`, `bits`, `uniform`, `randint` all behave the same at ~185–190 µs/iter),
and is unchanged by `JAX_MPS_ASYNC_DISPATCH=1`.

## Evidence it's the threefry op specifically

I bisected a realistic Metropolis-Hastings loop body (RNG + spin flip + small RBM matmul +
accept/reject). All of these keep the loop on the fast path at <20 µs/iter:

| loop body op                         | µs/iter |
|--------------------------------------|--------:|
| matmul + `tanh` (baseline)           |   ~6    |
| `shift_right_logical`, `bitwise_or`  |  ~13    |
| `rem` (integer modulo)               |  ~12    |
| `bitcast_convert_type` (u32→f32)     |  ~10    |
| `jnp.where`, `log(cosh)`, `one_hot`  |  ~15    |
| **`jax.random.bits` / `split` / `uniform` / `randint`** | **~185–190** |

So the individual integer/bit primitives that `uniform`/`randint` decompose into are all
fine — only the **threefry PRNG primitive** itself triggers the per-iteration synchronous
dispatch. (Note: a `jax.random.split` whose result is unused looks fast, but only because
XLA dead-code-eliminates it; once the key is threaded/used, the cost appears.)

`jax.make_jaxpr` shows `randint`/`uniform` contain **no `while`/`cond`** — there is no
nested control flow at the IR level — so this looks different from the "synchronous flush
when nested control flow exists" path mentioned in #194.

## Hypothesis on the underlying cause

The #194 fast path executes a detected counted loop without per-iteration condition
readback, flushing asynchronously. I suspect the **threefry2x32 primitive is lowered to a
path (custom call / non-fused kernel) that forces a synchronous device→host sync each time
it executes**, so when it appears in a loop body every iteration re-synchronizes — negating
the counted-loop optimization and the async-dispatch hiding. If threefry could be emitted
as a plain fused Metal kernel (it's pure integer arithmetic: adds, rotates, XORs — no host
interaction needed), the RNG loop should run at the same ~6 µs/iter as the arithmetic loop.

## Environment

- jax 0.10.2, jaxlib 0.10.2, **jax-mps 0.10.8**
- macOS (Apple Silicon, M-series)
- Also reproduced with `JAX_MPS_ASYNC_DISPATCH=1` (no change) and with `lax.scan` (same as `fori_loop`).

**Comment by PhilipVinc** [comment]
Another datapoint: using unsafe rbg actually leads to a large speedup

```python
import jax
jax.config.update("jax_default_prng_impl", "unsafe_rbg")  # <-- the workaround
```

but it would be interesting to fix it also on the default path.

**Comment by PhilipVinc** [comment]
A gentle ping, @tillahoffmann .
Apologises if you're on holiday.

**Comment by tillahoffmann** [comment]
Yes, was out last week. Thank you for the MWE, I'll try to reproduce here and have a look at a fix.

**Comment by tillahoffmann** [comment]
## Root cause, current state, and the plan to make threefry actually fast

This is a hand-off summary with enough context for a fresh session to pick up the remaining work.

### What the slowdown actually is (measured, not hypothesized)

The hypothesis in the issue description (a custom call / per-iteration sync) turned out to be wrong. Dumping the optimized IR (`JAX_MPS_DUMP_OPTIMIZED_IR`) shows the loop body contains **no custom call and no nested control flow** — `jax.random.split` inlines threefry2x32 as **~140 tiny elementwise StableHLO ops** per key op (20 rounds of add/xor/rotate on `tensor<2xui32>`, plus key pack/unpack).

The cost model is simple: inside a compiled loop body, **every MLX graph node that survives fusion costs ~1.5&nbsp;µs of dispatch** (M-series). Three factors multiplied together:

1. **Handler expansion** (fixed): the shift handlers expanded *every* shift into 7 MLX nodes including an unfusable `Full` (`zeros_like`) for StableHLO's OOB-shift masking — and threefry has ~40 shifts per key op. `mx::compile` can only fuse unary/binary/ternary/broadcast primitives, so each shift also broke fusion.
2. **MLX fusion limits** (measured, not worth tuning): `max_compile_depth = 11` and `max_compile_arrays = 24` in `mlx/compile.cpp`. Rotate diamonds `(x<<k)|(x>>j)` burn ~3 depth levels each, and every rotation constant is a segment input. A sweep of `MLX_MAX_COMPILE_DEPTH` at 11/24/48/96 (via the new env-override patch, 3 interleaved reps, idle machine) showed **no benefit** — all within the 130–170&nbsp;µs noise band. The shape fuses badly no matter the budget.
3. **Inherently unfusable tail**: the slice/reshape/concatenate/iota key-packing ops.

### What is already done (two stacked PRs, branches `fix/control-flow-correctness` and `perf/shift-constant-fast-path`)

- Constant-shift fast path: splat-constant amounts in `[0, bit_width)` emit a single MLX shift op (all of threefry qualifies); dynamic amounts keep the masked path, rebuilt from scalar constants so every node fuses. Plus native `bitwise_invert` for `stablehlo.not`, and MLX patch 12 (`MLX_MAX_COMPILE_DEPTH`/`MLX_MAX_COMPILE_ARRAYS` env overrides, defaults unchanged).
- **Benchmark: the MWE went 190 → 137&nbsp;µs/iter** (arithmetic body baseline: ~5&nbsp;µs/iter). A 120-rotate loop body went 210 → 135&nbsp;µs/iter, now identical to an equivalent xor chain — shifts are no longer special.
- Important context: the denser fusion exposed a latent correctness bug (MLX fused segments exceeding Metal's 31-argument-buffer kernel limit → eval-time throw on a stream worker thread → silently swallowed → never-written output buffers). Fixed in the correctness PR (MLX patch 13 fuse-time budget guard + explicit `rethrow_pending_exception` probes + a GPU-stream sync fixing #195). **Any further RNG/fusion perf work must land on top of those fixes.**

### Why incremental fusion work is capped

~90 surviving kernels × ~1.5&nbsp;µs ≈ the 137&nbsp;µs we see. Fusion tuning has maybe 2× left; getting near the ~5&nbsp;µs arithmetic baseline requires reducing the *op count* by ~10×, not fusing the same ops better.

### Proposed approach: intercept `threefry2x32` as a primitive

All of `jax.random` bottoms out in a single JAX primitive, `threefry2x32_p` (`jax/_src/prng.py`). It reaches the plugin as 140 unrolled ops only because no `mps` platform lowering is registered — XLA GPU solves this exact problem with a `cu_threefry2x32` custom call. Plan:

1. **Python lowering hook** (in the jax-mps Python package, at plugin registration time):
   ```python
   from jax._src import prng
   from jax.interpreters import mlir
   mlir.register_lowering(prng.threefry2x32_p, _mps_threefry_lowering, platform="mps")
   ```
   emitting `stablehlo.custom_call @mps.threefry2x32(k0, k1, c0, c1) -> (o0, o1)` (flat `uint32` operands, mirroring the primitive's signature). Wrap the registration in `try/except` and fall back to doing nothing — jax internals are private API, and the generic inline expansion stays correct (just slower) if the import ever breaks.
2. **Runtime kernel**: a branch in `HandleCustomCall` (`src/pjrt_plugin/ops/control_flow.cc`) lowering `@mps.threefry2x32` to a single Metal kernel via `mlx::core::fast::metal_kernel` (already linked; see the existing fast-kernel usages). The kernel is standard threefry2x32-20: rotation schedule `{13,15,26,6}` / `{17,29,16,24}`, key schedule `k2 = k0 ^ k1 ^ 0x1BD11BDA`, injection every 4 rounds — all directly visible in the dumped IR of the current expansion. ~40 lines of Metal.
3. **Do NOT reuse MLX's own RNG** (`mlx::core::random::bits`): MLX uses a modified splittable threefry variant. We need **bit-exactness with JAX's CPU backend** — the whole test harness compares against CPU, and user seeds must reproduce across backends. Writing the standard algorithm makes bit-exactness true by construction.

Expected result: the loop body collapses from ~140 nodes to a custom call plus a handful of packing ops (~10 nodes), which should land `split`-in-loop in the **10–25&nbsp;µs/iter** range. It also speeds up *every* `jax.random` call outside loops (sampling-heavy MCMC forward passes), which is the workload behind this issue and #83.

### Details a fresh agent should watch for

- **Bit-exactness is the acceptance criterion**: the existing `tests/configs/random.py` configs compare MPS vs CPU exactly for integer outputs. Also test with `jax.config.update("jax_threefry_partitionable", True)` — it changes the surrounding lowering but bottoms out in the same primitive; add configs for both settings.
- The primitive's counter layout / odd-size handling (the `iota`/`concatenate` logic in the current expansion) lives in JAX's `threefry2x32` wrapper *around* the primitive, so the custom call itself is a pure `(k0,k1,c0,c1) → (o0,o1)` map over equal-shaped arrays — keep the kernel that simple.
- Per project rules: register an `OperationTestConfig` for tests (no standalone test files), and add a `FusionTestConfig`-style IR check if a pass is involved (here it isn't — the custom call comes from the Python lowering, so `tests/test_fusion.py` machinery is not needed; a config exercising `jax.random.split/bits/uniform/randint` under jit suffices, most of which already exist).
- Benchmark before/after with the MWE from this issue; debugging aids: `JAX_MPS_DUMP_OPTIMIZED_IR=<dir>`, `JAX_MPS_NO_OPTIMIZE=1`, `MLX_MAX_COMPILE_DEPTH` (patch 12).
- If `mlx::core::fast::metal_kernel`'s C++ surface turns out to be awkward for multi-output kernels, the fallback implementation is two launches (one per output) or packing `(o0,o1)` as one `uint32[2,...]` array split afterwards — still ~2–3 nodes instead of 140.

**Comment by PhilipVinc** [comment]
For what is worth, I do not think that bit-exactness is strictly needed (I don't think that Jax generates bit-exact random numbers across different platforms like cpu and gpu?) but it might be handy to recycle Jax's test so if it carries no performance penalty they can be reused.
